static-site/src/components/ListResources/Showcase.tsx

```diff
@@ -46,9 +47,15 @@ export const Showcase = ({cards, setSelectedFilterOptions}: ShowcaseProps) => {
     <div>
       <ShowcaseContainer className={!isExpandable ? "" : isExpanded ? "expanded" : "collapsed"} $expandedHeight={cardsContainerHeight}>
         <CardsContainer ref={cardsContainerRef}>
-          {cards.map((el) => (
-            <ShowcaseTile card={el} key={el.name} setSelectedFilterOptions={setSelectedFilterOptions}/>
-            ))}
+          {cards.map((el: FilterCard | UrlCard) => {
+            if (Object.hasOwn(el, "filters")) {
```

Comment:
> the dangling return is unnecessary since you know `el` is one of those two types

the nice thing about the discriminated union way of doing this is that this `if` can be replaced by a switch statement, and you get compiler level verification of the exhaustiveness of the switch (i.e., the compiler yells if there's an un-covered case).

Reply:
> the `in` operator will narrow types for you, e.g. the following avoids the casts:

With the rework done in 70b94a6748a0abc3b7d3cc112ae1508f3d0ab31c I've changed it to simply condition on `el.filter`/`el.url`: b6e9a1d8a3ada5da4617513064a99ca2f1d2e654

Without the rework, the `in` operator might be better¹ but it wouldn't prevent the need for type casts which are necessary for the downstream component to realize it's accepting the right type.

¹ it's slightly [different](https://stackoverflow.com/a/13633027)

Reply:
I have questions here, but categorically none of this is blocking or a request for changes.

> Without the rework, the in operator might be better¹ but it wouldn't prevent the need for type casts which are necessary for the downstream component to realize it's accepting the right type.

Which type casts? With my proposed changes (on top of 644ea281) I get rid of the casts and don't have any errors from `tsc`.

>  I'd argue that this is assumption is fragile to future changes - it's possible that there will be other types in the future.

Isn't this precisely one of the reasons why we use TS tho? If we add an extra type`el: FilterCard | UrlCard | SomeOtherCard` then with my proposed changes I get a type error¹ because `UrlShowcaseTile`'s types aren't satisfied.


¹ Only if the new type doesn't define `url` because TS is structurally typed, but that's the same as your explicit checking for the `url` property.

Reply:
> Without the rework, the `in` operator might be better¹ but it wouldn't prevent the need for type casts which are necessary for the downstream component to realize it's accepting the right type.

One more time, then I will stop beating this :drum: …

The most idiomatic way of handling this (IMO) is the discriminated union approach. It's widely used in Typescript because it provides an easy way to narrow a type such that the compiler can ensure is exhaustively complete over the potential set of available types.

Reply:
@genehack see https://github.com/nextstrain/nextstrain.org/pull/901#discussion_r1633661373 for discriminated union discussion

Reply:
> @genehack see [#901 (comment)](https://github.com/nextstrain/nextstrain.org/pull/901#discussion_r1633661373) for discriminated union discussion

Yep, saw that. Disagree that it's not worth figuring out how to inject a `type` field; the value of knowing that this type of code is exhaustive is one of the big things Typescript brings to the table:

```typescript
const card = <either UrlCard or FilterCard>;
switch(card.type) {
  case "filter": ...
  case "url" ...
  // note: no default
}
```

whenever a new type of Card gets added, :point-up: will fail to compile and whoever is adding it will realize they need to update it. 

IMO, that's worth a lot.

Reply:
Ok, I'm open to reconsider adding a `type` field if the need arises again.